openstack/02a-MAClow/IEEE802154_security.c

```diff
@@ -0,0 +1,905 @@
+                                                          msg->l2_keyIndex,
+                                                          &msg->l2_keySource,
+                                                          idmanager_getMyID(ADDR_PANID),
+                                                          msg->l2_frameType);
+
+   if (keyDescriptor==NULL){
+      //can't find the key
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                           (errorparameter_t)msg->l2_frameType,
+                           (errorparameter_t)2);
+      return E_FAIL;
+   }
+
+   //g device descriptor lookup
+   deviceDescriptor = IEEE802154_security_deviceDescriptorLookup(&msg->l2_keySource,
+                                                                idmanager_getMyID(ADDR_PANID),
+                                                                keyDescriptor);
+
+   if (deviceDescriptor==NULL){
+      //can't find the device
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                           (errorparameter_t)msg->l2_frameType,
+                           (errorparameter_t)3);
+      return E_FAIL;
+   }
+
+   //h Security Level lookup
+   securityLevelDescriptor = IEEE802154_security_securityLevelDescriptorLookup(msg->l2_frameType,
+                                                                              msg->commandFrameIdentifier);
+
+   if (securityLevelDescriptor == NULL){
+	      //can't find the frame type
+	      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+	                           (errorparameter_t)msg->l2_frameType,
+	                           (errorparameter_t)4);
+	      return E_FAIL;
+   }
+
+   //i+j+k
+   if (IEEE802154_security_incomingSecurityLevelChecking(securityLevelDescriptor,
+                                                        msg->l2_securityLevel,
+                                                        deviceDescriptor->Exempt) ==FALSE) 
+      {
+      //security level not allowed
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                           (errorparameter_t)msg->l2_frameType,
+                           (errorparameter_t)5);
+      return E_FAIL;
+   }
+
+   //l+m Anti-Replay - not needed for TSCH mode
+
+   //n Control of key used
+   if (IEEE802154_security_incomingKeyUsagePolicyChecking(keyDescriptor,
+                                                         msg->l2_frameType,
+                                                         0)==FALSE){
+     // improper key used
+     openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                          (errorparameter_t)msg->l2_frameType,
+                          (errorparameter_t)6);
+     return E_FAIL;
+   }
+
+   //create nonce
+   memset(&nonce[0], 0, 13);
+   switch (msg->l2_keyIdMode){
+      case 0:
+      case 1:
+         for (i=0; i<8; i++){
+            nonce[i] = ieee802154_security_vars.m_macDefaultKeySource.addr_64b[i];
```

Comment:
same remark for nonce creation here, you should use extended/short address from source address field in the frame..